rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -449,22 +467,64 @@ extern "C" rmw_ret_t rmw_set_log_severity(rmw_log_severity_t severity)
   return RMW_RET_OK;
 }
 
+static void dds_listener_callback(dds_entity_t entity, void * arg)
+{
+  // Currently we don't use the entity which calls this callback.
+  // We need the dds_entity_t arg just to comply with the function pointer
+  // signature of 'dds_on_data_on_readers_fn' defined on dds_public_listener.h
+  (void)entity;
+
+  // Cast the listener callback arg to user callback data.
+  struct user_callback_data_t * data;
+  data = static_cast<user_callback_data_t *>(arg);
+
+  // Call the user callback
+  data->callback(
+    data->user_data,
+    {data->entity_handle, data->event_type});
+}
+
 extern "C" rmw_ret_t rmw_subscription_set_listener_callback(
   rmw_subscription_t * rmw_subscription,
   rmw_listener_callback_t callback,
   const void * user_data,
   const void * subscription_handle)
 {
-  (void)rmw_subscription;
-  (void)callback;
-  (void)user_data;
-  (void)subscription_handle;
-  // auto subscription = static_cast<CddsSubscription *>(rmw_subscription->data);
-  // subscription->setCallback(user_data, callback, subscription_handle);
-  RCUTILS_LOG_ERROR_NAMED(
-    "rmw_node.cpp",
-    "rmw_subscription_set_listener_callback: not supported (yet)");
-  return RMW_RET_UNSUPPORTED;
+  auto sub = static_cast<CddsSubscription *>(rmw_subscription->data);
+  auto entity_to_listen = sub->enth;
+
+  // This API sets two callbacks:
+  // 1. The DDS subscription's listener callback.
+  // 2. The user callback, called from the listener callback.
+  /*
+      dds_listener_callback(..) {
+        user_callback(..);
+      }
+  */
+
+  // Set the user callback data (if valid)
+  user_callback_data_t * data = &(sub->user_callback_data);
+
+  if (user_data && subscription_handle && callback) {
+    data->callback = callback;
```

Comment:
we should always update `sub->user_callback_data` regardless of whether they are valid or null pointers.
otherwise you will have dangling memory stored there.

for example you want to unset the callbacks and destroy an executor, but here there is still a pointer to the executor

Reply:
the whole function could be re-organized a little bit to make it easier to understand.

```
user_callback_data_t * data = &(sub->user_callback_data);

data->callback = callback;
data->user_data = user_data;
data->event_type = SUBSCRIPTION_EVENT;
data->entity_handle = subscription_handle;

if (user_data && subscription_handle && callback) {
    sub->listener = dds_create_listener(data);
    dds_lset_data_on_readers(sub->listener, dds_listener_callback);
    return dds_set_listener(entity_to_listen, sub->listener);
} else {
    return dds_set_listener(entity_to_listen, NULL);
}

```

Reply:
Moreover, I'm not sure about the `if (user_data && subscription_handle && callback) {` check.

Should it be with OR instead of AND ?

Then the events executor should check whether pointers are valid or not when it receives stuff.

Reply:
Regarding comment 1 and 2, OK.
Coment 3, I think is OK to check the 3 pointers are valid here:

1. This is not directly related to events executor, the callback could be something different in the future, so checks have to be done here.
2. The `user_data`, `subscription_handle` and `callback` are used in `dds_listener_callback`, so if I do:
`if (user_data || subscription_handle ||  callback)` , maybe `user_data` and `subscription_handle` are valid but `callback == nullptr` which will cause a seg fault when `dds_listener_callback` is called
3. Checking valid stuff in the events executor every time something is pushed could bring performance degradations, so better check them all here, only once

Reply:
sorry, my suggestion about the OR was wrong.

What we should do here is to only check if `callback` is not a null pointer.
That's the only item that is used in the RMW layer, while the others are simply forwarded.

These callbacks are not the events executor callback, they are generic functions that are planned to be used for different use-cases.
Some use-cases will definitely have a null `user_data`.
In theory a callback could also not have any `subscription_handle` (I think they always should, but it's a different story).

The events executor should always check that the pointer it receives are valid.
Checking a couple of pointers to be valid will not cause any performance degradation.
They come from a completely different, RMW-specific, repository, so there is no control on what is put there.

Reply:
Okoko sounds better checking just the callback.
Regarding `user_data` should be check on `push_event` API of the `EventsExecutor`
```
  static void
  push_event(const void * executor_ptr, rmw_listener_event_t event)
  {
    if(!executor_ptr) return; // Or throw a runtime error
 ```
and `<entity>_handle` is already checked on
```
EventsExecutorEntitiesCollector::get_subscription(const void * subscription_id)
{
  auto it = weak_subscriptions_map_.find(subscription_id); 
  // If not found, returns nullptr
 ```